metals/src/main/scala/scala/meta/internal/metals/codeactions/FlatMapToForComprehensionCodeAction.scala

```diff
@@ -0,0 +1,503 @@
+      document <- buffers.get(path)
+      termApply <- maybeTree
+      termApplyArg <- termApply.args.headOption
+      indentation = getIndentForPos(termApplyArg.pos, document)
+    } yield codeActionWithApply(
+      path,
+      termApply,
+      indentation
+    )
+    maybeChainedCodeAction.flatten.toSeq
+
+  }
+
+  private def constructCodeAction(
+      forElementsList: List[Enumerator],
+      yieldTerm: Term,
+      indentation: String,
+      path: AbsolutePath,
+      startPos: l.Position,
+      endPos: l.Position
+  ): l.CodeAction = {
+
+    val indentedElems = forElementsList
+      .map(
+        _.syntax
+          .split(Array('\n'))
+          .map(line => s"$indentation   $line")
+          .mkString("\n")
+      )
+      .mkString("\n")
+
+    val yieldTermIndentedString = yieldTerm.syntax
+      .split(Array('\n'))
+      .map(line => s"$indentation   $line")
+      .mkString("\n")
+
+    val forYieldString =
+      s"""|{
+          |$indentation for {
+          |$indentedElems
+          |$indentation }  yield {
+          |$yieldTermIndentedString
+          |$indentation }
+          |$indentation}""".stripMargin
+
+    val codeAction = new l.CodeAction()
+    val range =
+      new l.Range(startPos, endPos)
+    codeAction.setTitle(
+      FlatMapToForComprehensionCodeAction.flatMapToForComprehension
+    )
+    codeAction.setKind(this.kind)
+    val forComprehensionTextEdit = new l.TextEdit(range, forYieldString)
+    codeAction.setEdit(
+      new l.WorkspaceEdit(
+        Map(
+          path.toURI.toString -> List(forComprehensionTextEdit).asJava
+        ).asJava
+      )
+    )
+    codeAction
+  }
+
+  private def findTopMostApply(
+      currentTermApply: Term.Apply,
+      lastValidTermApply: Option[Term.Apply]
+  ): Option[Term.Apply] = {
+
+    currentTermApply.fun match {
+      case term if term.isNot[Term.Select] => lastValidTermApply
+      case termSelect: Term.Select
+          if termSelect.name.value != "map" && termSelect.name.value != "flatMap" &&
+            termSelect.name.value != "filter" && termSelect.name.value != "withFilter" && termSelect.name.value != "filterNot" =>
+        lastValidTermApply
+      case _: Term.Select =>
+        if (
+          !currentTermApply.parent.exists(parent =>
+            parent.is[Term.Select] && parent.parent.exists(_.is[Term.Apply])
+          )
+        ) {
+          Some(currentTermApply)
+        } else {
+          val result = Some(currentTermApply)
+          val grandParentApply =
+            currentTermApply.parent.get.parent.get.asInstanceOf[Term.Apply]
+          findTopMostApply(grandParentApply, result)
+        }
+    }
+  }
+
+  private def codeActionWithApply(
+      path: AbsolutePath,
+      termApply: Term.Apply,
+      indentation: String
+  ): Option[l.CodeAction] = {
+
+    findTopMostApply(termApply, None).flatMap { topMostApply =>
+      val nameGenerator = MetalsNames(topMostApply, "generatedByMetals")
```

Comment:
Can we make the generated intermediate value names shorter like `x`?

```suggestion
      val nameGenerator = MetalsNames(topMostApply, "x")
```

Reply:
well, the rationale behind it was to avoid shading the variables that might already exist in the outer context as much as possible. So it is far less likely for someone to have an awkward variable name such as `generatedByMetals0` in their pre-existing code, than let’s say `newName0` or `x0`